cores/esp8266/core_esp8266_spi_utils.cpp

```diff
@@ -0,0 +1,144 @@
+#include "spi_flash.h"
+
+// for PRECACHE_*
+#include "core_esp8266_features.h"
+
+#include "spi_utils.h"
+
+extern "C" uint32_t Wait_SPI_Idle(SpiFlashChip *fc);
+
+/*
+ * critical part of SPICommand.
+ * Kept in a separate function to aid with precaching
+ * PRECACHE_* saves having to make the function IRAM_ATTR.
+ *
+ * Note: if porting to ESP32 mosi/miso bits are set in 2 registers, not 1.
+ */
+static SpiOpResult PRECACHE_ATTR
+_SPICommand(volatile uint32_t spiIfNum,
+            uint32_t spic,uint32_t spiu,uint32_t spiu1,uint32_t spiu2,
+            uint32_t *data,uint32_t write_words,uint32_t read_words)
+{ 
+  if (spiIfNum>1) return SPI_RESULT_ERR;
+
+  // force SPI register access via base+offest. 
+  // Prevents loading individual address constants from flash.
+  uint32_t *spibase = (uint32_t*)(spiIfNum ? &(SPI1CMD) : &(SPI0CMD));
+  #define SPIREG(reg) *((volatile uint32_t *)(spibase+(&(reg) - &(SPI0CMD))))
+
+  // preload any constants and functions we need into variables
+  // Everything must be volatile or the optimizer can treat them as 
+  // constants, resulting in the flash reads we're trying to avoid
+  void    *(* volatile memcpyp)(void *,const void *, size_t) = memcpy;
+  uint32_t (* volatile Wait_SPI_Idlep)(SpiFlashChip *) = Wait_SPI_Idle;
+  volatile SpiFlashChip *fchip=flashchip;
+  volatile uint32_t spicmdusr=SPICMDUSR;
+
+  if (!spiIfNum) {
+     // Only need to precache when using SPI0
+     PRECACHE_START();
+     Wait_SPI_Idlep((SpiFlashChip *)fchip);
+  }
+  
+  uint32_t old_spi_usr = SPIREG(SPI0U);
+  uint32_t old_spi_usr2= SPIREG(SPI0U2);
+  uint32_t old_spi_c   = SPIREG(SPI0C);
+
+  //SPI0S &= ~(SPISE|SPISBE|SPISSE|SPISCD);
+  SPIREG(SPI0C) = spic;
+  SPIREG(SPI0U) = spiu;
+  SPIREG(SPI0U1)= spiu1;
+  SPIREG(SPI0U2)= spiu2;
+
+  if (write_words>0) {
+     // copy the outgoing data to the SPI hardware
+     memcpyp((void*)&(SPIREG(SPI0W0)),data,write_words*4);
+  }
+
+  // Start the transfer
+  SPIREG(SPI0CMD) = spicmdusr;
+
+  // wait for the command to complete
+  uint32_t timeout = 1000;
+  while ((SPIREG(SPI0CMD) & spicmdusr) && timeout--) {}
+
+  if ((read_words>0) && (timeout>0)) {
+     // copy the response back to the buffer
+     memcpyp(data,(void *)&(SPIREG(SPI0W0)),read_words*4);
+  }
+
+  SPIREG(SPI0U) = old_spi_usr;
+  SPIREG(SPI0U2)= old_spi_usr2;
+  SPIREG(SPI0C) = old_spi_c;
+  
+  PRECACHE_END();
+  return (timeout>0 ? SPI_RESULT_OK : SPI_RESULT_TIMEOUT);
+}
+
+
+/*  SPI0Command: send a custom SPI command.
+ *  This part calculates register values and passes them to _SPI0Command().
+ */
+SpiOpResult SPI0Command(uint8_t cmd, uint32_t *data, uint32_t mosi_bits, uint32_t miso_bits) {
+  if (mosi_bits>(64*8)) return SPI_RESULT_ERR;
+  if (miso_bits>(64*8)) return SPI_RESULT_ERR;
+
+  uint32_t mosi_words=mosi_bits/32;
+  uint32_t miso_words=miso_bits/32;
+  if (mosi_bits % 32 != 0) mosi_words++;
+  if (miso_bits % 32 != 0) miso_words++;
+
+  uint32_t spiu=SPIUCOMMAND; //SPI_USR_COMMAND
```

Comment:
From here until the call to _SPICommsnd, I can't understand anything of the code. Please add comments, e. g. reminders of the meanings of the registers, what the operations mean, etc.